### ISSUE TITLE
[Node] Seeing different .endDialog behaviour with LuisDialog as parent 

### ISSUE THREAD

**Issue by iangriggs** [Microsoft/BotBuilder#304]
I've a scenario where I'm using a LuisDialog as an initial classifier that performs a .beginDialog to a separate child route running a secondary dialog.

My understanding was that when the secondary dialog either explicitly called .endDialog or it naturally comes to the end of its 'waterfall' flow it would end the secondary dialog execution and return control back to the initial LuisDialog. 

After debugging I found that the flow of control goes back to the LuisDialog but it has a stored state that includes the 'intent' of the secondary dialog it results in immediately jumps back into the secondary dialog. The appearance is that I seem stuck in the secondary dialog with no way to get back to the first. I know that I can call session.reset to force myself back to the start but that seems a sledgehammer approach and not how you intend things to work.
 
As a test I tried going from a simple waterfall primary/initial dialog through to the same secondary dialog and found the behaviour I expected in that control is returned back to the primary dialog as expected.

As its difficult to explain clearly I've taken a copy of the hello-Textbox example and fashioned a repro demonstrating both the LuisDialog and a simple waterfall as described above. I've attached this repro to this issue. Comment/Uncomment the different scenarios in the app.js and interact with the TextBox conversation and hopefully it demonstrates the different in operation.

[hello-LuisIssue.zip](https://github.com/Microsoft/BotBuilder/files/268516/hello-LuisIssue.zip)

Many thanks in advance for looking at this.

**Comment by Stevenic** [comment]
Waterfalls don't automatically end a dialog. You would need to call `endDialog()` as the last step of your waterfall. The reason for that is that it's kid of gray knowing when a wataerfall is completed. You have to call another dialog to actually pump the watertfall.

**Comment by Stevenic** [comment]
So for one in your `luisDialog.onBegin()` handler you're missing the `next` param which you need to call to continue processing the dialog. So it shouldn't even be calling LUIS.

The core issue is you're falling into a trap I set for you. When your handler is a simple closure it will get called twice. Once when the handler is invoked and a second time when the dialog you called completes. This second call is causing you to call back in a loop to the other dialog.  The samples I created have protection code for this but i'll admit it's not clear. I actually added a warning in the docs because everyone keeps hitting this.

I'm actually planning a break change for v1.0 that will fix this. You can avoid this by always using a waterfall so when I make this change a simple closure will just be the first step of a waterfall.  For now you can just always use waterfalls.

```

luisDialog.on('Help', [
    function(session) {
        session.send('\'Help\' intent fired');
        session.beginDialog('/secondary');
    }
])
```

**Comment by Stevenic** [comment]
By the way... There are times when you want the current behavior. If you were building a new custom prompt from scratch for instance. I'm planning to expose an internal class called SimpleDialog which is what I use to host a simple closure today. That way if you want the old behavior you can just wrap your call with a SimpleDialog.

**Comment by iangriggs** [comment]
Thanks for taking the time to respond so comprehensively. 

I updated the repro sample I previously attached above with the waterfall rather than a simple closure and, as you suggested, it worked perfectly and paused on the primary/first dialog waiting for input...will use that technique going forward.

Just for completeness in following the information you provided above I tried adding the `next()` call in the `luisDialog.onBegin()` but that seemed to have a negative effect and resulted in the 'Help' intent firing all of its own accord before I even typed anything. Quite likely I've misunderstood what you meant and not a problem as not including a `next()` seems work fine and results in the flow waiting to receive a keyboard entry and only then firing the relevant intent handler (the single 'Help' one in my sample code).

This was the my `.onBegin` btw. based on what I thought you were suggesting:
```
luisDialog.onBegin(function(session, args, next){
    session.send('On begin - say \'help\' to fire intent'); 
    next();    // <-- added this but actually not having it seems to work better
});
```
Feel free to close this issue unless there is anything more to say on the `next()` in the `.onBegin`. I only referenced the `.next()` situation out of interest.

Thanks again for your explaination. I'll be keeping an eye on the commits relating ot the v1.0 update you alluded to above.
